File: backend/src/routes/early-access.ts
import { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify'
import { z } from 'zod'
import { prisma } from '../services/database.service'
import { sendEarlyAccessEmail } from '../services/email.service'

const TURNSTILE_SECRET_KEY = process.env.TURNSTILE_SECRET_KEY || '0x4AAAAAACCXwEK23TURNS62usztaHWOSnE'

async function verifyTurnstileToken(token: string, ip?: string): Promise<boolean> {
  // Skip verification if using test key
  if (TURNSTILE_SECRET_KEY.startsWith('1x0000')) {
    return true
  }

  const response = await fetch('https://challenges.cloudflare.com/turnstile/v0/siteverify', {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      secret: TURNSTILE_SECRET_KEY,
      response: token,
      ...(ip && { remoteip: ip }),
    }),
  })

  const data = await response.json() as { success: boolean }
  return data.success
}

const earlyAccessSchema = z.object({
  email: z.string().email(),
  company: z.string().min(2),
  currentSolution: z
    .enum(['kafka', 'dms', 'fivetran', 'none', 'other'])
    .optional(),
  dataVolume: z.enum(['<10GB', '10-100GB', '100GB-1TB', '>1TB']).optional(),
  message: z.string().optional(),
  turnstileToken: z.string().min(1, 'Captcha verification required'),
})

type EarlyAccessBody = z.infer<typeof earlyAccessSchema>

export async function earlyAccessRoutes(fastify: FastifyInstance) {
  fastify.post<{ Body: EarlyAccessBody }>(
    '/api/early-access',
    {
      schema: {
        body: {
          type: 'object',
          required: ['email', 'company', 'turnstileToken'],
          properties: {
            email: { type: 'string', format: 'email' },
            company: { type: 'string', minLength: 2 },
            currentSolution: {
              type: 'string',
              enum: ['kafka', 'dms', 'fivetran', 'none', 'other'],
            },
            dataVolume: {
              type: 'string',
              enum: ['<10GB', '10-100GB', '100GB-1TB', '>1TB'],
            },
            message: { type: 'string' },
            turnstileToken: { type: 'string' },
          },
        },
      },
    },
    async (request: FastifyRequest<{ Body: EarlyAccessBody }>, reply: FastifyReply) => {
      try {
        // Validate with Zod
        const data = earlyAccessSchema.parse(request.body)

        // Verify Turnstile token
        const isValidCaptcha = await verifyTurnstileToken(data.turnstileToken, request.ip)
        if (!isValidCaptcha) {
          return reply.status(400).send({
            success: false,
            message: 'Captcha verification failed',
          })
        }

        // Save to database
        const earlyAccessRequest = await prisma.earlyAccessRequest.create({
          data: {
            email: data.email,
            company: data.company,
            currentSolution: data.currentSolution,
            dataVolume: data.dataVolume,
            message: data.message,
            ipAddress: request.ip,
            userAgent: request.headers['user-agent'],
          },
        })

        // Send notification email (async, don't wait)
        sendEarlyAccessEmail(data).catch((err) =>
          fastify.log.error(err, 'Failed to send email')
        )

        fastify.log.info(
          { requestId: earlyAccessRequest.id, email: data.email },
          'New early access request'
        )

        return reply.status(201).send({
          success: true,
          message: 'Request submitted successfully',
        })
      } catch (error) {
        if (error instanceof z.ZodError) {
          return reply.status(400).send({
            success: false,
            errors: error.errors.map((e) => ({
              path: e.path.join('.'),
              message: e.message,
            })),
          })
        }

        fastify.log.error(error, 'Early access request failed')
        return reply.status(500).send({
          success: false,
          message: 'Internal server error',
        })
      }
    }
  )
}

File: backend/internal/handlers/early_access.go
package handlers

import (
	"encoding/json"
	"net/http"
	"net/url"

	"github.com/savegress/platform/backend/internal/services"
)

// EarlyAccessHandler handles early access requests
type EarlyAccessHandler struct {
	earlyAccessService *services.EarlyAccessService
	turnstileSecret    string
}

// NewEarlyAccessHandler creates a new early access handler
func NewEarlyAccessHandler(earlyAccessService *services.EarlyAccessService, turnstileSecret string) *EarlyAccessHandler {
	return &EarlyAccessHandler{
		earlyAccessService: earlyAccessService,
		turnstileSecret:    turnstileSecret,
	}
}

// EarlyAccessRequest represents the early access form data
type EarlyAccessRequest struct {
	Email           string `json:"email"`
	Company         string `json:"company"`
	CurrentSolution string `json:"currentSolution,omitempty"`
	DataVolume      string `json:"dataVolume,omitempty"`
	Message         string `json:"message,omitempty"`
	TurnstileToken  string `json:"turnstileToken"`
}

// Submit handles early access form submissions
func (h *EarlyAccessHandler) Submit(w http.ResponseWriter, r *http.Request) {
	var req EarlyAccessRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	// Validate required fields
	if req.Email == "" || req.Company == "" {
		respondError(w, http.StatusBadRequest, "email and company are required")
		return
	}

	if req.TurnstileToken == "" {
		respondError(w, http.StatusBadRequest, "captcha verification required")
		return
	}

	// Verify Turnstile token
	if !h.verifyTurnstile(req.TurnstileToken, getClientIP(r)) {
		respondError(w, http.StatusBadRequest, "captcha verification failed")
		return
	}

	// Get client info
	ipAddress := getClientIP(r)
	userAgent := r.UserAgent()

	// Save to database
	err := h.earlyAccessService.Submit(r.Context(), services.EarlyAccessInput{
		Email:           req.Email,
		Company:         req.Company,
		CurrentSolution: req.CurrentSolution,
		DataVolume:      req.DataVolume,
		Message:         req.Message,
		IPAddress:       ipAddress,
		UserAgent:       userAgent,
	})
	if err != nil {
		respondError(w, http.StatusInternalServerError, "failed to submit request")
		return
	}

	respondCreated(w, map[string]interface{}{
		"success": true,
		"message": "Request submitted successfully",
	})
}

func (h *EarlyAccessHandler) verifyTurnstile(token, ip string) bool {
	// Skip verification for test keys
	if len(h.turnstileSecret) > 6 && h.turnstileSecret[:6] == "1x0000" {
		return true
	}

	data := url.Values{}
	data.Set("secret", h.turnstileSecret)
	data.Set("response", token)
	if ip != "" {
		data.Set("remoteip", ip)
	}

	resp, err := http.PostForm("https://challenges.cloudflare.com/turnstile/v0/siteverify", data)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	var result struct {
		Success bool `json:"success"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return false
	}

	return result.Success
}

func getClientIP(r *http.Request) string {
	// Check X-Forwarded-For header first (for proxies)
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		return xff
	}
	if xri := r.Header.Get("X-Real-IP"); xri != "" {
		return xri
	}
	return r.RemoteAddr
}
